feat: add documentation on verification via code (#1120)
ory/docs: docs/kratos/self-service/flows/verify-email-account-activation.mdx

import AjaxWarning from "@theme/AjaxWarning"
import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
```

> Ory allows users to verify the addresses of their account. This is important to prove that the user has access to the address they

Author: Done.

import AjaxWarning from "@theme/AjaxWarning"
import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
```

Ory allows users to verify email addresses associated with their accounts. This is important to prove that the user has access to
the address they used to create their account. If verification is enabled, Ory Identities starts the verification process
automatically. A user can also verify their address manually.

The verification flow is supported in both browsers and API clients.

:::caution

> Email verification should not be taken as a guarantee that the person currently using the account, is the person that verified the
> address. It is recommended to implement additional security mechanisms.

Author: ```suggestion
Completing account verification does not guarantee that the account is used by the person who performed verification. We recommend implementing additional security mechanisms to ensure that verified accounts aren't taken over by malicious actors.
```

What about this?

> address. It is recommended to implement additional security mechanisms.

Reviewer: That's good! I edited your suggestion to turn `does not` into `doesn't`. Contractions are cool and ok to use!

> address. It is recommended to implement additional security mechanisms.

Author: fine with me then

import AjaxWarning from "@theme/AjaxWarning"
import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"

> import CodeLinkComparison from "./partials/code-link-comparison.mdx"

Reviewer: That's not how we do it, please update for both docs.
Also, the imports are not referenced in the documents.

https://github.com/ory/docs#markdown-partials

Author: Done, my bad.

 }
```

### Email Templates

Ory Identities (Ory Kratos) comes with default email templates for verification flows.

You can quickly replace the defaults and customize the messages to match the look and feel of your solution. Read the
[custom email template documentation](../../emails-sms/05_custom-email-templates.mdx) to learn more.

:::note

When users attempt to initiate verification for unregistered addresses, the system sends a message to the email address that was
used in the attempt.

This prevents account enumeration attacks as explained in this
[blog post by Troy Hunt](https://www.troyhunt.com/website-enumeration-insanity-how-our-personal-data-is-leaked/).

:::

### Require a verified address for login

Depending on your use case, you might want to limit what a user without at least one verified address is able to do in your

> application. Ory Identities supports requiring a verified address before the user can log into their account.

Reviewer: The way you put it made me initially think that there's more you can do than just preventing login.

Author: Yup, was copied from the old doc, but doesn't make much sense. I removed it.

> Depending on your use case, you might want to limit what a user without at least one verified address is able to do in your
> application. Ory Identities supports requiring a verified address before the user can log into their account.

Reviewer: ```suggestion
Ory Identities supports allowing only users with a verified email address to sign in.
```
I think this is enough.

Author: Done.

```mdx-code-block
<Tabs groupId="oc-or-oss">
  <TabItem value="ocp" label="The Ory Network" default>
```

> Go to the **Authentication** → **Email Verification** page in the [Ory Console](https://console.ory.sh/guides) and enable **
> Require Verified Address for Login**.

Reviewer: ```suggestion
Go to **Authentication** → **Email Verification** in the [Ory Console](https://console.ory.sh/guides) and enable **Require Verified Address for Login**.
```

Author: Done.

```mdx-code-block
  </TabItem>

>   <TabItem value="oss" label="Ory Identities">

Reviewer: This shouldn't be there I think?

Author: Hmm, So how would you propose the structure then?

```

Adjust your Ory Identities (Ory Kratos) configuration as follows:

```diff
selfservice:
  flows:
    login:
+      after:
+        hooks:
+        - hook: require_verified_address
```

For more information see the [hooks configuration](../../hooks/01_configure-hooks.mdx) documentation.

```mdx-code-block
  </TabItem>
</Tabs>
```

:::note

Please be aware, that when requiring a verified address for login, a typo in the email address set by the user either during the

> registration or when changing their email address, will make the login for that user impossible.

Reviewer: I wanted to re-write it but a question came to my mind - is this really a problem? 
I mean, I created an account for `tmoek@email.com` and I must verify it before I do anything. 
I can't verify bc my mail is `tomek@email.com`. 
I just create a new account for a mail w/o typo and verify. 
Am I missing anything?

Author: Eh, I guess you're right. This was in the old doc, but it should be implied. I'll remove the note.

:::

## Choosing the right strategy

Ory supports two strategies for verifying your user's addresses.

```mdx-code-block
<Tabs>
  <TabItem value="code" label="One-time codes">
```

> When using one-time codes to verify a user's address, Ory Identities sends out an email containing a 6-digit code. If the user has
> access to the mailbox they can enter the code they received to verify their address.

Reviewer: ```suggestion
When using one-time codes to verify a user's address, Ory Identities sends out an email containing a 6-digit code. The user must enter the code in a dedicated UI text field to verify their address.
```

Author: Done.

Alternatively, the email also contains a link to the form that pre-fills the code if the user clicks on it. The user must still

> submit the form for Ory Identities to actually mark the address as verified.

Reviewer: So how does it work? I click the link and the form opens, with the code pre-filled in the text box? All I have to do is press "submit"?

Author: yes

```mdx-code-block
  </TabItem>
  <TabItem value="oss" label="Magic Links">
```

> When choosing to send magic links to verify a user's address, Ory Identities sends out an email containing a "magic link". If the
> user has access to the mailbox of that account and clicks the link, Ory Identities marks that address as verified.

Reviewer: ```suggestion
When using the "magic link" strategy, Ory Identities sends the user an email with a "magic link". If the
user has access to the email associated with the and clicks the link, Ory Identities marks that address as verified.
```

Reviewer: I used "strategy", but shouldn't it be "method" as mentioned earlier?

Author: Done.

email associated with the account and clicks the link, Ory Identities marks that address as verified.

```mdx-code-block
  </TabItem>
</Tabs>
```

If the user completes the verification after the link or code expired, they must provide the email address associated with their
account to get a new verification email.

### Comparison

```mdx-code-block
import CodeLinkComparison from "./_common/code-link-comparison.mdx"

<CodeLinkComparison />
```

## Code examples

The user interface for account verification is a page in your solution that renders the actual form elements for the user.

In contrast to other identity systems, Ory Identities (Ory Kratos) doesn't render this HTML directly. Instead, you need to
implement the HTML code in your solution, which gives you complete flexibility and customizability in your user interface flows
and designs. This part of your application then directly interfaces with Ory Identities through the API.

The API responds with a JSON document describing the form elements to render and actions the form should take upon submission,
cancellation, etc. The following shows examples for a few different languages and frameworks.

> <RenderFlow flow="verification" />

Reviewer: This is weird - that "Code examples" thing is now text only. The `<RenderFlow flow="verification" />` doesn't seem to do anything. Can you see what's going on here?

Author: Whoops, removed the import a bit prematurely, but the tooling didn't complain so I didn't notice. Good catch.

Go to **Authentication** → **Email Verification** in the [Ory Console](https://console.ory.sh/projects/current/verification) and
enable **Require Verified Address for Login**.

```mdx-code-block
  </TabItem>
  <TabItem value="cli" label="Ory CLI">
```

1. Download the Ory Identities config from your project and save it to a file:

   ```shell
   ## List all available projects
   ory list projects

   ## Get config
   ory get identity-config {project-id} --format yaml > identity-config.yaml
   ```

2. Add the configuration for the verification flow

   ```diff
   selfservice:
     flows:
       login:
   +      after:
   +        hooks:
   +        - hook: require_verified_address
   ```

3. Update the Ory Identities configuration using the file you worked with:

   ```shell
   ory update identity-config {project-id} --file updated_config.yaml
   ```

:::tip

For more information see the [hooks configuration](../../hooks/01_configure-hooks.mdx) documentation.

:::

```mdx-code-block
  </TabItem>
</Tabs>
```

## Choosing the right strategy

Ory supports two strategies for verifying your user's addresses.

```mdx-code-block
<Tabs>
  <TabItem value="code" label="One-time codes">
```

> When using one-time codes to verify a user's address, Ory Identities sends out an email containing a 6-digit code. The user must
> enter the code in a dedicated UI text field to verify their address.

Reviewer: ```suggestion
When using one-time codes to verify a user's address, Ory Identities sends an email with a 6-digit code to the user. They must
enter the code in a dedicated UI text field to verify their address.
```

> Alternatively, the email also contains a link to the form that pre-fills the code if the user clicks on it. The user must still
> submit the form for Ory Identities to actually mark the address as verified.

Reviewer: ```suggestion
The email also contains a link, which takes the user to the verification form with the code pre-filled in the appropriate UI field. The only thing the user must do to verify their address is to submit the form. 
```

Author: done, not sure why `Now, the user only has to submit the form to complete the verification.` isn't good though....

```mdx-code-block
  </TabItem>
  <TabItem value="link" label="Magic Links">
```

> When using the "magic link" method, Ory Identities sends the user an email with a "magic link". If the user has access to the
> email associated with the and clicks the link, Ory Identities marks that address as verified.

Reviewer: ```suggestion
When using the "magic link" method, Ory Identities sends the user an email with a link that serves as the means of verification. If the user has access to the
email associated with the account and clicks the link, Ory Identities marks that address as verified.
```

Author: Not really sure if that sentence ("that serves as the means of verification") adds anything more, TBH. Would you be okay with leaving it out?

Reviewer: "that serves as the means of verification" is an explanation of what is that "magic link" that we refer to. By reading that the user knows that the "magic link" is a link sent for the purpose of verification. Saying that the "magic link methods sens magic links in emails" is too vague. What is the link for in that case? 
But maybe it'd work better if it were shorter? If you have an idea how to do that then sure, please show me :)

Author: Yeah, but that's explained in the sentence afterwards, no?

Author: When using the "magic link" method, Ory Identities sends the user an email with a **verification** link. If the user has access to the
email associated with the account and clicks the link, Ory Identities marks that address as verified.

What about this? (bold for highlighting)

Reviewer: That sounds good! Go ahead and use it. No bolding, of course ;)
https://github.com/ory/docs#bolding-code-formatting

```mdx-code-block
  </TabItem>
</Tabs>
```

> If the user completes the verification after the link or code expired, they are asked to re-enter their email address and are sent
> a new email for verification.

Reviewer: ```suggestion
If the user completes the verification after the link or code expired, 
they must provide the email address associated with their account to get a new verification email.
```

Author: Done.

### Comparison

"One-time code" (`code`) is the default, preferred, and recommended method by Ory.

Ory supports "magic links", but considers this method a legacy solution. Currently, the method is supported but should be
considered deprecated and is set to be removed in future releases.

Consider using the `code` method as it mitigates many of the drawbacks of "magic links":

- Some email virus scanners open links in emails to scan them. This invalidates the link and may prevent users from recovering
  their account even if they have the access to the defined recovery address.
- Verification flows initialized by apps on mobile phones or smart devices don't work with links.
- Depending on the device settings, clicking a link from an email can open a different browser than the one used to initialize the

>   verification flow. This can confuse your users.

Reviewer: I see that this content is both in this doc and in another one in this PR. Is this necessary?

Author: It's slightly different.

Author: I've extracted it now. PTAL.

Reviewer: https://github.com/ory/docs/pull/1120#discussion_r1042337805

:::caution

When you change the verification strategy from `link` to `code` in an existing project, you might need to adjust your UI. Make
sure the flow works correctly with your UI implementation after changing the recovery strategy.

:::

## Code examples

> The user interface for the account verification is a page in your solution that should render the actual form elements for the
> user.

Reviewer: ```suggestion
The user interface for account verification is a page in your solution that renders the actual form elements for the user.
```

   ```shell
   ## List all available projects
   ory list projects

   ## Get config
   ory get identity-config {project-id} --format yaml > identity-config.yaml
   ```

2. Add the configuration for the verification flow

   ```diff
   selfservice:
     flows:
       login:
   +      after:
   +        hooks:
   +        - hook: require_verified_address
   ```

3. Update the Ory Identities configuration using the file you worked with:

   ```shell
   ory update identity-config {project-id} --file updated_config.yaml
   ```

:::tip

For more information see the [hooks configuration](../../hooks/01_configure-hooks.mdx) documentation.

:::

```mdx-code-block
  </TabItem>
</Tabs>
```

## Choosing the right strategy

Ory supports two strategies for verifying your user's addresses.

```mdx-code-block
<Tabs>
  <TabItem value="code" label="One-time codes">
```

When using one-time codes to verify a user's address, Ory Identities sends an email with a 6-digit code to the user. They must
enter the code in a dedicated UI text field to verify their address.

The email also contains a link, that takes the user to the verification form with the code pre-filled in the appropriate form

> field. Now, the user only has to submit the form to complete the verification.

Author: Done.
